### PR TITLE
sign-to-contract blog post

### PR DESCRIPTION
**Pull request by LeoComandini**
Blog post describing elliptic curve commitments, *pay-to-contract*, *sign-to-contract* and the integration of *sign-to-contract* with the Electrum timestamp plugin.

### CODE REVIEW

File: _posts/2018-04-11-sign-to-contract.md [hunk 1]
```diff
@@ -0,0 +1,400 @@
...
+  The signature is committed in the `wtxid`,
+  but not in the `txid`: 
+  the latter is committed in the transaction Merkle tree, 
+  while the former is committed in another Merkle tree, 
+  whose tip is inserted in the coinbase transaction. 
+  As a result, the signature is committed in the block header, 
+  but the path has to traverse both trees and the coinbase.
+  Thus proofs double in size 
+  and the miner may decide to include arbitrary data in the proofs, 
+  since he has some control on what is inside the coinbase.
```
Comment (petertodd) [inline _posts/2018-04-11-sign-to-contract.md]: This should say that proofs *at least* double in size: if the coinbase happens to be large they may be significantly larger.

Also, if the coinbase is really large - multiple KB - it won't be possible to make a OTS proof from it as OTS has a maximum message size within the proof processing. Which also could be an issue if you try to make a timestamp with a really big transaction (not likely to be a problem in your Electrum plugin).

File: _posts/2018-04-11-sign-to-contract.md [hunk 2]
```diff
@@ -0,0 +1,400 @@
...
+    r = R.x mod n
+    s = k^(-1) * (m + r*x) mod n
+    return (r,s)
+```
+Carl wants to spend the coins locked in a given address whose private key is `x`.
+He composes an unsigned transaction `uTX` sending from that address to Diana's.
+To sign the transaction he applies `ECDSAsign(x,m)` where `m` is the hash value of `uTX` conveniently serialized.
+Once produced the signature `(r,s)` Carl fill `uTX` with an encoded version of the signature `SIG(r,s)`, 
+obtaining the signed (and broadcastable) transaction `TX`.
+Then sends `TX` to the network and, later on, when confirmed (~6 blocks), it becomes part of the blockchain forever.
```
Comment (petertodd) [inline _posts/2018-04-11-sign-to-contract.md]: Let's remove the "(~6 blocks)" bit: there's no specific number of confirmations that makes a transaction "confirmed". The Satoshi whitepaper just used 6 in an example; it's not a magic number.

File: _posts/2018-04-11-sign-to-contract.md [hunk 3]
```diff
@@ -0,0 +1,400 @@
...
+- an actual implementation w\ Electrum
+    - how to run
+    - how to timestamp
+    - how to verify
+- conclusions
+
+# A bit of history
+When hearing the name *sign-to-contract*, 
+you may wonder from where the contract comes out.
+To find an answer we have to go back (in 2012) to one of the [first applications](https://arxiv.org/pdf/1212.3257.pdf) that was developed: *pay-to-contract* (*sign-to-contract* elder brother) which basically allows to embed a contract inside a receiver public key.
```
Comment (petertodd) [inline _posts/2018-04-11-sign-to-contract.md]: No, pay-to-contract *commits* to a contract in the public key; it doesn't *embed* the contract.

File: _posts/2018-04-11-sign-to-contract.md [hunk 4]
```diff
@@ -0,0 +1,400 @@
...
+When a customers performs a payment to a merchant, with *pay-to-contract*, 
+he can bind the bill containing the details of the purchase to the receiving address.
```
Comment (petertodd) [inline _posts/2018-04-11-sign-to-contract.md]: Again, say "commit" rather than "bind"

File: _posts/2018-04-11-sign-to-contract.md [hunk 5]
```diff
@@ -0,0 +1,400 @@
...
+
```
Comment (petertodd) [inline _posts/2018-04-11-sign-to-contract.md]: I think the article would be more clear if you put a description of what a secp256k1 commitment actually is here, and then explained how it gets used in transactions to commit to things. Right now it's not clear what it actually is.

File: _posts/2018-04-11-sign-to-contract.md [hunk 6]
```diff
@@ -0,0 +1,400 @@
...
+...
+sha256
+verify BitcoinBlockHeaderAttestation(...)
+# Bitcoin merkle root ...
+```
+So Alice has created an OTS timestamp for her document without spending a single satoshi.
+
+But, wait, has Alice received the payment from Bob?
+
+She has, but the question is delicate.
```
Comment (petertodd) [inline _posts/2018-04-11-sign-to-contract.md]: I think a better way to explain the above is to just say that Alice has created a new pubkey, but it's derived based on information that she hasn't necessarily backed up already; there's no question that she's been paid.

File: _posts/2018-04-11-sign-to-contract.md [hunk 7]
```diff
@@ -0,0 +1,400 @@
...
+The value obtained is an elliptic curve point that serves two (independent) purposes: 
+is a public key __and__ is a commitment to `(m,P)` 
+(or only to `m`, if `P` is given before).
+
+To timestamp we need to write the x-coord of the elliptic curve point `C(m,P)` on the blockchain, 
+there are two places where such points can be found, leading to two techniques:
+- Address/Public key: *pay-to-contract*
+- Signature: *sign-to-contract*
+
+## *pay-to-contract*
```
Comment (petertodd) [inline _posts/2018-04-11-sign-to-contract.md]: It's not clear to me that you actually need this section at all: it's of historical interest, and not relevant as an intro to timestamping with secp256k1 commitments.

File: _posts/2018-04-11-sign-to-contract.md [hunk 8]
```diff
@@ -0,0 +1,400 @@
...
+if, in that period, Carl secret key `x` gets exposed, an eavesdropper, Eve, will redirect Carl's coin to her. 
+However, once the transaction to Diana is confirmed, Eve cannot do much (unless Carl does address reuse), 
+since the coins locked by `x` are already spent.
+The private key can also be revealed in a indirect way, 
+in fact in a signature revealing the nonce `k` 
+(or `e` to be consistent with the above formula) 
+is equivalent to expose `x`.
+Still *sign-to-contract* does not increase the risks Carl faces with a standard signature.
+Indeed when showing his OTS proof he just claims that what can be found  in the signature (namely `Q.x`) was obtained from the elliptic curve point (`Q=e*G`) which is the sum of two other specific points (`R` and `h(R||c)*G`),
+but actually this does not give any additional clue to solve the discrete logarithm that secures the signature.
```
Comment (petertodd) [inline _posts/2018-04-11-sign-to-contract.md]: I think this part is confusing to the reader. You're talking about secret key compromise in a way that doesn't make it clear that it's not an issue.

File: _posts/2018-04-11-sign-to-contract.md [hunk 9]
```diff
@@ -0,0 +1,400 @@
...
+
+Another way to expose `x` is signing two times with two different nonces
+and then spreading both signatures;
+this may happen if Carl uses *sign-to-contract* with 
+`(x,m,c1)` and `(x,m,c2)` and then broadcasts both transactions.
+He should take care not to do so.
```
Comment (petertodd) [inline _posts/2018-04-11-sign-to-contract.md]: How is this an issue? Signing the same message twice should be is fine.

File: _posts/2018-04-11-sign-to-contract.md [hunk 10]
```diff
@@ -0,0 +1,400 @@
...
+
+In addition, if Carl looses the contract along with its hash `c`, 
+then he looses the ability of proving the timestamp,
+but he does not compromise any coin: 
+his coins have been already spent and Diana has hers in the desired destination, 
+which is independent from the contract.
+For this reason *sign-to-contract* should be preferred when timestamping.
+
+## Remarks
+- The name contract is merely due to historical reasons.
```
Comment (petertodd) [inline _posts/2018-04-11-sign-to-contract.md]: You could avoid this terminology issue by - like I mentioned above - introducing the whole concept as a secp256k1 *commitment*, and if you want, noting how historically it came out of a desire to *commit* to contracts when paying someone.

**Review by petertodd — CHANGES_REQUESTED** [review]

File: _posts/2018-04-11-sign-to-contract.md (hunk 9 above)
Comment (LeoComandini) [reply _posts/2018-04-11-sign-to-contract.md]: That was confusing, I took it out.
I was thinking about the case with two signatures for `(x,m,c1)` and `(x,m,c2)` made with a deterministic nonce `k`. If the two singatures and both `c1`, `c2` and `R=kG` are exposed, then there are two linear equations for the second part of the signatures, but only two secrets `k` and `x`.
However that is a very particular case and stating it there is confusing.

File: _posts/2018-04-11-sign-to-contract.md (hunk 10 above)
Comment (LeoComandini) [reply _posts/2018-04-11-sign-to-contract.md]: Many thanks for all the comments, I should have fixed all of those here 8d415e3bf5fbd2a9670e35cc4d3532feabc00d71.

**Pull request merged**